vic/drivers/classic/src/vic_populate_model_state.c

```diff
@@ -0,0 +1,101 @@
+ * contents are checked to see if it agrees with the current simulation set-up,
+ * if so it is used to initialize the model state.  If no state file is
+ * provided the model initializes all variables with defaults and the user
+ * should expect to throw out the beginning of the simulation period as model
+ * start-up.
+ *
+ * @section LICENSE
+ *
+ * The Variable Infiltration Capacity (VIC) macroscale hydrological model
+ * Copyright (C) 2014 The Land Surface Hydrology Group, Department of Civil
+ * and Environmental Engineering, University of Washington.
+ *
+ * The VIC model is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *****************************************************************************/
+
+#include <vic_driver_classic.h>
+
+/******************************************************************************
+ * @brief    Initialize the model state (energy balance, water balance, and
+ *           snow components).
+ *****************************************************************************/
+void
+vic_populate_model_state(all_vars_struct *all_vars,
+                         filep_struct     filep,
+                         size_t           cellnum,
+                         soil_con_struct *soil_con,
+                         veg_con_struct  *veg_con,
+                         lake_con_struct  lake_con)
+{
+    extern option_struct options;
+
+    size_t               Nveg;
+    int                  tmp_lake_idx;
+
+    cell_data_struct   **cell;
+    energy_bal_struct  **energy;
+    lake_var_struct     *lake;
+    snow_data_struct   **snow;
+    veg_var_struct     **veg_var;
+
+    cell = all_vars->cell;
+    energy = all_vars->energy;
+    lake = &all_vars->lake_var;
+    snow = all_vars->snow;
+    veg_var = all_vars->veg_var;
+
+    Nveg = veg_con[0].vegetat_type_num;
+
+    // Initialize all data structures to 0
+    initialize_soil(cell, soil_con, Nveg);
+    initialize_snow(snow, Nveg);
+    initialize_veg(veg_var, Nveg);
+    if (options.LAKES) {
+        tmp_lake_idx = lake_con.lake_idx;
+        if (tmp_lake_idx < 0) {
+            tmp_lake_idx = 0;
+        }
+        initialize_lake(lake, lake_con, soil_con, &(cell[tmp_lake_idx][0]),
+                        false);
+    }
+    initialize_energy(energy, Nveg);
+
```

Comment:
@yixinmao - I can't say what the cause of the inexact restarts may be.  But to my knowledge, the derived variables are not saved in the state file.  All state variables are initialized to 0 in the `initialize_*` functions (e.g., `initialize_soil()`).  Then, some are overwritten by either reading the state file or generating the default state, and others are subsequently overwritten by `create_derived*` (this is the same set of variables regardless of whether a state file was read or a default state was set).  So if there are any uninitialized variables, they simply need to be added to the `initialize*` functions.

Yes, the set of variables that are assigned values in `generate_default*` is smaller than the set of variables stored in the state files.  This is because only those variables that need to have non-zero values are set in `generate_default*`.

There is one main exception to this: there are many lake state variables saved in the state file that are actually redundant, i.e. can be derived from the others.  There already is an open issue about this but I think it is slated for 5.1 as it is not essential to getting 5.0 running correctly.